service/src/main/kotlin/fi/espoo/evaka/nekku/NekkuQueries.kt

```diff
@@ -0,0 +1,87 @@
+// SPDX-FileCopyrightText: 2017-2025 City of Espoo
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+package fi.espoo.evaka.nekku
+
+import fi.espoo.evaka.decision.logger
+import fi.espoo.evaka.shared.db.Database
+import org.jdbi.v3.core.mapper.PropagateNull
+
+data class CustomerNumbers(
+    @PropagateNull val number: String,
+    val name: String,
+    val customergroup: String,
+    val unit_size: String,
+)
+
+/** Throws an IllegalStateException if Nekku returns an empty texture list. */
+fun fetchAndUpdateNekkuCustomers(
+    client: NekkuClient,
+    db: Database.Connection,
+    warner: (s: String) -> Unit = loggerWarner,
+) {
+    val customersFromNekku =
+        client.getCustomers().map { it ->
+            CustomerNumbers(it.number, it.name, it.customergroup, it.unit_size)
+        }
+
+    if (customersFromNekku.isEmpty())
+        error("Refusing to sync empty Nekku customer list into database")
+    db.transaction { tx ->
+        val nulledCustomersCount =
+            tx.resetNekkuCustomerNumbersNotContainedWithin(customersFromNekku)
+        if (nulledCustomersCount != 0)
+            warner(
+                "Nekku custoner list update caused $nulledCustomersCount customer numbers to be set to null"
+            )
+        val deletedCustomerCount = tx.setCustomerNumbers(customersFromNekku)
+        logger.info {
+            "Deleted: $deletedCustomerCount Nekku customer numbers, inserted ${customersFromNekku.size}"
+        }
+    }
+}
+
+fun Database.Transaction.resetNekkuCustomerNumbersNotContainedWithin(
+    nekkuCustomerNumbers: List<CustomerNumbers>
+): Int {
+    val newNekkuCustomerNumbers = nekkuCustomerNumbers.map { it.number }
+    val affectedRows = execute {
+        sql(
+            "UPDATE daycare_group SET nekku_customer_number = null WHERE nekku_customer_number != ALL (${bind(newNekkuCustomerNumbers)})"
+        )
+    }
+    return affectedRows
+}
+
+fun Database.Transaction.setCustomerNumbers(customerNumbers: List<CustomerNumbers>): Int {
+    val newCustomerNumbers = customerNumbers.map { it.number }
+    val deletedCustomerCount = execute {
+        sql("DELETE FROM nekku_customer WHERE number != ALL (${bind(newCustomerNumbers)})")
+    }
+    executeBatch(customerNumbers) {
+        sql(
+            """
+INSERT INTO nekku_customer (number, name, customer_group, unit_size)
+VALUES (
+    ${bind{it.number}},
+    ${bind{it.name}},
+    ${bind{it.customergroup}},
+    ${bind{it.unit_size}}
+)
+ON CONFLICT (number) DO UPDATE SET
+  name = excluded.name
```

Comment:
Tässä tulisi varmaan päivittää nimen lisäksi myös `customer_group` ja `unit_size`. Muuten synkronointi ei koskaan päivitä olemassaoleville `number`eille näitä tietoja.